docs/source/configure.rst
.. _configuration_toplevel:

.. _configuring_oyProjectManager:

Configuring oyProjectManager
============================

To configure oyProjectManager and make it fit to your studios need you should
use the ``config.py`` file as mentioned in next sections.

config.py File
--------------

oyProjectManager uses the ``config.py`` to let one to customize the system
config.

The ``config.py`` file is searched in a couple of places through the
system:
    
  * under "~/.oypmrc/" directory (not yet)
  * under "$OYPROJECTMANAGER_PATH"

The first path is a folder in the users home dir. The second one is a path
defined by the ``OYPROJECTMANAGER_PATH`` environment variable.

Defining the ``config.py`` by using the environment variable gives the most
customizable and consistent setup through the studio. You can set
``OYPROJECTMANAGER_PATH`` to a shared folder in your fileserver where all
the users can access.

Because, ``config.py`` is a regular Python code which is executed by
oyProjectManager, you can do anything you were doing in a normal Python
script. This is very handy if you have another source of information which
is reachable by a Python script.

In config.py you can:
  
  * Add new users
  * Customize default project structure
  * Customize file naming convention
  * Customize path naming convention
  * Customize the database file name
  * Customize server paths
  * Customize environments (host programs running oyProjectManager)
  * and many more

If there is no ``OYPROJECTMANAGER_PATH`` variable in your current
environment or it is not showing an existing path or there is no
``config.py`` file the system will use the system defaults and it is
probably not going to work as desired.

Config Variables
----------------

Variables which can be set in ``config.py`` are as follows:

.. confval:: asset_thumbnail_filename
   
   A Jinja2 template showing the filename of
   :class:`~oyProjectManager.models.asset.Asset` thumbnails. The default
   value is::
     
     asset_thumbnail_filename = "{{asset.code}}_thumbnail.{{extension}}"

.. confval:: asset_thumbnail_path
   
   A Jinja2 template showing the path of
   :class:`~oyProjectManager.models.asset.Asset` thumbnails. The default
   value is::
   
     asset_thumbnail_path = "{{project.code}}/Assets/{{asset.type}}/{{asset.code}}/Thumbnail"

.. confval:: database_url
   
   The URL of the database the default value is::
   
     "sqlite:///$OYPROJECTMANAGER_PATH/project_manager.db"
  
   in which the database is a SQLite3 file based database and it is placed
   right beside the config path.

.. confval:: default_asset_type_name
   
   The default type name for newly created
   :class:`~oyProjectManager.models.asset.Asset` instances, not used if
   another value is passed with the ``type`` argument. The default value is
   "Generic".

.. confval:: default_fps
   
   The default fps value for newly created
   :class:`~oyProjectManager.models.project.Project` instances, not used if
   another value is supplied with the ``fps`` argument. The default value
   is "25".

.. confval:: default_resolution_preset
   
   The name of the default resolution preset used in newly created
   :class:`~oyProjectManager.models.project.Project` instances. Not used if
   the ``width``, ``height`` and ``aspect_ratio`` arguments are used. The
   default value is "HD 1080", the default value should exists in
   :confval:`resolution_presets` config value.

.. confval:: default_take_name
   
   The default take name value for newly created
   :class:`~oyProjectManager.models.version.Version` instances, not used if
   another value is supplied with the ``take`` argument. The default value
   is "MAIN".

.. confval:: environments
   
   A list of dictionaries holding environments info. Environments are the
   host programs (Maya, Houdini, Nuke etc.) that oyProjectManager is
   running on. This config value holds the name and the native extensions
   of that environment. The default value is::
   
     environments = [
        {
            "name":"Maya",
            "extensions":["ma", "mb"]
        },
        {
            "name":"Houdini",
            "extensions":["hip"]
        },
        {
            "name":"Nuke",
            "extensions": ["nk"],
        },
        {
            "name":"Photoshop",
            "extensions": ["psd", "pdd"],
            "export_extensions": ["tif", "tga", "bmp", "jpg", "iff"],
        },
        {
            "name":"3DEqualizer",
            "extensions": ["3te"]
        },
        {
            "name":"Fusion",
            "extensions": ["comp"]
        }
     ]

.. confval:: file_size_format
   
   The string formatting used in version file size info columns in UI. The
   default value is::
   
     file_size_format = '%.2f MB'

.. confval:: project_structure
   
   The default project structure template for newly created
   :class:`~oyProjectManager.models.project.Project` instances. It is a
   Jinja2 template where every line is a definition of a folder in the
   project structure. The default value is::
     
     project_structure = '''{% for sequence in project.sequences %}
         {% set seq_path = project.full_path + '/Sequences/' + sequence.code %}
         {{seq_path}}/Edit/Offline
         {{seq_path}}/Edit/Sound
         {{seq_path}}/References/Artworks
         {{seq_path}}/References/Text/Scenario
         {{seq_path}}/References/Photos_Images
         {{seq_path}}/References/Videos
         {{seq_path}}/References/Others
         {% for shot in sequence.shots %}
             {{seq_path}}/Shots/{{shot.code}}
             {{seq_path}}/Shots/{{shot.code}}/Plate
             {{seq_path}}/Shots/{{shot.code}}/Ref
             {{seq_path}}/Shots/{{shot.code}}/Texture
         {% endfor %}
     {% endfor %}
     '''
   
   oyProjectManager will supply the ``project`` variable to the Jinja2
   template engine.

.. confval:: repository_env_key
   
   The name of the environment variable showing the repository path. The
   default value is "REPO". If you change this, you should also update the
   environment variables of all the computers using oyProjectManager with
   that value.

.. confval:: resolution_presets
   
   A dictionary holding resolution presets. This info is used while
   creating :class:`~oyProjectManager.models.project.Project` instances. The
   default value is::
   
      resolution_presets = {
        "PC Video": [640, 480, 1.0],
        "NTSC": [720, 486, 0.91],
        "NTSC 16:9": [720, 486, 1.21],
        "PAL": [720, 576, 1.067],
        "PAL 16:9": [720, 576, 1.46],
        "HD 720": [1280, 720, 1.0],
        "HD 1080": [1920, 1080, 1.0],
        "1K Super 35": [1024, 778, 1.0],
        "2K Super 35": [2048, 1556, 1.0],
        "4K Super 35": [4096, 3112, 1.0],
        "A4 Portrait": [2480, 3508, 1.0],
        "A4 Landscape": [3508, 2480, 1.0],
        "A3 Portrait": [3508, 4960, 1.0],
        "A3 Landscape": [4960, 3508, 1.0],
        "A2 Portrait": [4960, 7016, 1.0],
        "A2 Landscape": [7016, 4960, 1.0],
        "50x70cm Poster Portrait": [5905, 8268, 1.0],
        "50x70cm Poster Landscape": [8268, 5905, 1.0],
        "70x100cm Poster Portrait": [8268, 11810, 1.0],
        "70x100cm Poster Landscape": [11810, 8268, 1.0],
        "1k Square": [1024, 1024, 1.0],
        "2k Square": [2048, 2048, 1.0],
        "3k Square": [3072, 3072, 1.0],
        "4k Square": [4096, 4096, 1.0],
      }

.. confval:: rev_number_padding
   
   The amount of padding applied to the
   :attr:`~oyProjectManager.models.version.Version.revision_number` attribute
   of :class:`~oyProjectManager.models.version.Version` class while generating
   filename of that Version instance. In default filename and path
   templates the revision_number attribute is not used. The default value
   is "2".

.. confval:: rev_number_prefix
   
   The prefix of
   :attr:`~oyProjectManager.models.version.Version.revision_number` attribute
   of :class:`~oyProjectManager.models.version.Version` class while generating
   filename of that Version instance. In default filename and path
   templates the revision_number attribute is not used. The default value
   is "r".

.. confval:: sequence_format
   
   oyProjectManager uses PySeq, a Python package to compress and uncompress

CHANGELOG
Version History
===============

0.2.5.3
-------

* Fixed mayaEnv.Maya, it was trying to convert None to absolute path when
  checking for external paths.

* Added post_open() to houdiniEnv.Houdini.

* In version_updater the column label "Latest" is changed to
  "Latest Published" to make it clear.

* Fixed mayaEnv.Maya.export_as(), the workspace.mel is now properly created
  upon "export as".

0.2.5.2
-------

* mayaEnv.Maya now creates all the folders defined in the
  pymel.core.workspace.fileRules dictionary.

* Creating a new Asset in version_creator now trigger the recreation of the
  project structure, this will allow any asset related folders to be created
  automatically. Change ``project_structure`` config value as desired.

* Added a new config value called ``maya_workspace_file_content`` to hold the
  workspace.mel content.

* version_creator UI now can work in read-only mode where it is only used for
  choosing existing versions.

0.2.5.1
-------

* Deletes are now cascaded. When a Project instance is deleted, all the related
  Sequence, Assets are also deleted, which also deletes the Shots and all the
  Version instances related to the deleted Assets and Shots. Also deleting a
  Sequence will delete the related Shots and thus the related Versions, and
  deleting an Asset will delete all the related Versions.

* In version_creator UI the notes of previous versions can now be changed by
  using the context menu on the previous_versions_tableWidget.

* project_properties dialog now warns the user for empty code field

* project_properties dialog now updates the code field when the name field is
  changed.

* mayaEnv now warns the user and rejects saving the file if there are external
  references (file textures, mentalray textures, ibl nodes etc.), which are not
  under $REPO.

* In version_creator the take name is now formatted for newly created takes, so
  every first letter of every word is now uppercase (title) and any fancy
  characters including empty spaces are replaced with underscores. And the
  newly added take name is selected in the list, and also the list is now
  sorted after adding a new take name, and duplicate take names are now
  prevented.

* Updated the version_updater, it now correctly displays the take of the
  versionable and also the 'Update' checkbox is now working correctly.

* version_creator now lists only active users.

0.2.5
-----

* Moved back all the database classes to their own modules and removed the
  oyProjectManager.core.models module. All the classes are going to be imported
  to ``oyProjectManager`` namespace, thus it will easy again to import them as
  if they are in the same module::
    
    # instead of
    from oyProjectManager.core.models import Project, Sequence, Shot, Asset
    
    # you can do import them like this.
    from oyProjectManager import Project, Sequence, Shot, Asset

* Tests are now using in-memory database, which is way faster then the the file
  based version.

* Added the FileLink class to hold links to files in the file system. The
  Version class will use this class in later versions. But for now it is used
  in IOMixin.

* Added the first ever mixin called IOMixin. It adds two new attributes to the
  mixed in class called ``inputs`` and ``outputs``. The inputs and outputs
  attributes are holding list of FileLink instances.

* Version class is now mixed in with IOMixin so it accepts FileLink instances
  in its inputs and outputs attributes.

* "Version Creator" UI is now able to run in environmentless mode in which it
  will only generate a new version but only copy the path to the clipboard
  instead of telling the environment to save the document to a specific path.

* Added two new context menu to the previous_versions_tableWidget_context_menu
  first one called "Copy Path" and the second "Copy Output Path", where in the
  first menu users will be able to copy the specified versions full_path and in

oyProjectManager/config.py
# -*- coding: utf-8 -*-
# Copyright (c) 2009-2012, Erkan Ozgur Yilmaz
# 
# This module is part of oyProjectManager and is released under the BSD 2
# License: http://www.opensource.org/licenses/BSD-2-Clause

import os
import logging

logger = logging.getLogger(__name__)

class Config(object):
    """Config abstraction
    
    Idea is coming from Sphinx config.
    
    Holds system wide configuration variables. See
    `configuring oyProjectManager`_ for more detail.
    
    .. _configuring oyProjectManager: ../configure.html
    """
    
    default_config_values = dict(
        
        database_url = "sqlite:///$OYPROJECTMANAGER_PATH/project_manager.db",
        
        status_list = [
            'WTS',
            'WIP',
            'REV',
            'APP',
            'CMP'
        ],
        
        status_list_long_names = [
            'Waiting To Start',
            'Work In Progress',
            'For Review',
            'Approved',
            'Completed'
        ],
        
        status_bg_colors = [
            (192,  80,  77), #WTS
            (255, 192,   0), #WIP
            ( 89, 141, 213), #REV
            (155, 187,  89), #APP
            (155, 187,  89), #CMP
        ],
        
        status_fg_colors = [
            (255, 255, 255), #WTS
            (  0,   0,   0), #WIP
            (  0,   0,   0), #REV
            (  0,   0,   0), #APP
            (  0,   0,   0), #CMP
        ],
        
        sequence_format = "%h%p%t %R",
        
        shot_number_prefix = "SH",
        shot_number_padding = 3,
        
        rev_number_prefix = "r",
        rev_number_padding = 2,
        
        ver_number_prefix = "v",
        ver_number_padding = 3,
        
        default_fps = 25,
        
        default_asset_type_name = "Generic",
        default_take_name = "Main",
        
        users_data = [{"name": "Administrator", "initials": "adm"}],
        
        # just use one repository for now
        repository_env_key = "REPO",
        
        repository = {
            "name": "Default",
            "windows_path": "~/Projects",
            "linux_path": "~/Projects",
            "osx_path": "~/Projects"
        },


        file_size_format = "%.2f MB",
        time_format = '%d.%m.%Y %H:%M',

        environments = [
            {
                "name":"Maya",
                "extensions":["ma", "mb"]
            },
            {
                "name":"Houdini",
                "extensions":["hip"]
            },
            {
                "name":"Nuke",
                "extensions": ["nk"],
            },
            {
                "name":"Photoshop",
                "extensions": ["psd", "pdd"],
                "export_extensions": ["tif", "tga", "bmp", "jpg", "iff"],
            },
            {
                "name":"3DEqualizer",
                "extensions": ["3te"]
            },
            {
                "name":"Fusion",
                "extensions": ["comp"]
            }
        ],
        
        resolution_presets = {
            "PC Video": [640, 480, 1.0],
            "NTSC": [720, 486, 0.91],
            "NTSC 16:9": [720, 486, 1.21],
            "PAL": [720, 576, 1.067],
            "PAL 16:9": [720, 576, 1.46],
            "HD 720": [1280, 720, 1.0],
            "HD 1080": [1920, 1080, 1.0],
            "1K Super 35": [1024, 778, 1.0],
            "2K Super 35": [2048, 1556, 1.0],
            "4K Super 35": [4096, 3112, 1.0],
            "A4 Portrait": [2480, 3508, 1.0],
            "A4 Landscape": [3508, 2480, 1.0],
            "A3 Portrait": [3508, 4960, 1.0],
            "A3 Landscape": [4960, 3508, 1.0],
            "A2 Portrait": [4960, 7016, 1.0],
            "A2 Landscape": [7016, 4960, 1.0],
            "50x70cm Poster Portrait": [5905, 8268, 1.0],
            "50x70cm Poster Landscape": [8268, 5905, 1.0],
            "70x100cm Poster Portrait": [8268, 11810, 1.0],
            "70x100cm Poster Landscape": [11810, 8268, 1.0],
            "1k Square": [1024, 1024, 1.0],
            "2k Square": [2048, 2048, 1.0],
            "3k Square": [3072, 3072, 1.0],
            "4k Square": [4096, 4096, 1.0],
        },
        
        default_resolution_preset = "HD 1080",
        
        project_structure = """{% for sequence in project.sequences %}
            {% set seq_path = project.full_path + '/Sequences/' + sequence.code %}
            {{seq_path}}/Edit/Offline
            {{seq_path}}/Edit/Sound
            {{seq_path}}/References/Artworks
            {{seq_path}}/References/Text/Scenario
            {{seq_path}}/References/Text/Brief
            {{seq_path}}/References/Photos_Images
            {{seq_path}}/References/Videos
            {{seq_path}}/References/Others
            {{seq_path}}/References/Storyboard
            {% for shot in sequence.shots %}
                {{seq_path}}/Shots/{{shot.code}}
                {{seq_path}}/Shots/{{shot.code}}/Plate
                {{seq_path}}/Shots/{{shot.code}}/Reference
                {{seq_path}}/Shots/{{shot.code}}/Texture
            {% endfor %}
        {% endfor %}
        {% for asset in project.assets%}
            {% set asset_path = project.full_path + '/Assets/' + asset.type + '/' + asset.code %}
            {{asset_path}}/Texture
            {{asset_path}}/Reference
        {% endfor %}
        """,
        
        asset_thumbnail_path = "{{project.code}}/Assets/{{asset.type}}/{{asset.code}}/Thumbnail",
        asset_thumbnail_filename = "{{asset.code}}_thumbnail.{{extension}}",
        
        shot_thumbnail_path = "{{project.code}}/Sequences/{{sequence.code}}/Shots/{{shot.code}}/Thumbnail",
        shot_thumbnail_filename = "{{shot.code}}_thumbnail.{{extension}}",
        
        thumbnail_format = "jpg",
        thumbnail_quality = 70,
        thumbnail_size = [320, 180],
        
        version_types = [
            {
                "name": "Animation",
                "code": "Anim",
                "path": "{{project.code}}/Sequences/{{sequence.code}}/Shots/{{version.base_name}}/{{type.code}}",
                "filename": "{{version.base_name}}_{{version.take_name}}_{{type.code}}_v{{'%03d'|format(version.version_number)}}_{{version.created_by.initials}}{{version.extension}}",
                "output_path": "{{version._path}}/Output/{{version.take_name}}",
                "extra_folders": "",
                "environments": ["Maya", "Houdini"],
                "type_for": "Shot"
            },
            {
                "name": "Camera",
                "code": "Cam",
                "path": "{{project.code}}/Sequences/{{sequence.code}}/Shots/{{version.base_name}}/{{type.code}}",
                "filename": "{{version.base_name}}_{{version.take_name}}_{{type.code}}_v{{'%03d'|format(version.version_number)}}_{{version.created_by.initials}}{{version.extension}}",
                "output_path": "{{version._path}}/Output/{{version.take_name}}",
                "extra_folders": "",
                "environments": ["Maya", "Houdini"],
                "type_for": "Shot"
            },
            {
                "name": "Composition",
                "code": "Comp",
                "path": "{{project.code}}/Sequences/{{sequence.code}}/Shots/{{version.base_name}}/{{type.code}}",
                "filename": "{{version.base_name}}_{{version.take_name}}_{{type.code}}_v{{'%03d'|format(version.version_number)}}_{{version.created_by.initials}}{{version.extension}}",
                "output_path": "{{version._path}}/Output/{{version.take_name}}/v{{'%03d'|format(version.version_number)}}",
                "extra_folders": "",
                "environments": ["Nuke", "Fusion"],
                "type_for": "Shot"
            },
#            {
#                "name": "Edit",
#                "code": "Edit",
#                "path": "{{project.code}}/Sequences/{{sequence.code}}/Shots/{{version.base_name}}/{{type.code}}",
#                "filename": "{{version.base_name}}_{{version.take_name}}_{{type.code}}_v{{'%03d'|format(version.version_number)}}_{{version.created_by.initials}}{{version.extension}}",
#                "output_path": "{{version._path}}/Output/{{version.take_name}}",
#                "extra_folders": "",
#                "environments": ["Nuke", "Fusion"],
#                "type_for": "Shot"
#            },
            {
                "name":"FX",
                "code": "FX",
                "path": "{{project.code}}/Sequences/{{sequence.code}}/Shots/{{version.base_name}}/{{type.code}}",
                "filename": "{{version.base_name}}_{{version.take_name}}_{{type.code}}_v{{'%03d'|format(version.version_number)}}_{{version.created_by.initials}}{{version.extension}}",
                "output_path": "{{version._path}}/Output/{{version.take_name}}",
                "extra_folders": """{{version.path}}/anim
                        {{version.path}}/cache
                        {{version.path}}/exports""",
                "environments": ["Maya", "Houdini"],
                "type_for": "Shot"
            },
            {
                "name":"Model",
                "code": "Model",
                "path": "{{project.code}}/Assets/{{asset.type}}/{{version.base_name}}/{{type.code}}",
                "filename": "{{version.base_name}}_{{version.take_name}}_{{type.code}}_v{{'%03d'|format(version.version_number)}}_{{version.created_by.initials}}{{version.extension}}",
                "output_path": "{{version._path}}/Output/{{version.take_name}}",
                "extra_folders": "",
                "environments": ["Maya", "Houdini"],
                "type_for": "Asset"
            },
            {
                "name": "Other",
                "code": "Other",
                "path": "{{project.code}}/Assets/{{asset.type}}/{{version.base_name}}/{{type.code}}",
                "filename": "{{version.base_name}}_{{version.take_name}}_{{type.code}}_v{{'%03d'|format(version.version_number)}}_{{version.created_by.initials}}{{version.extension}}",
                "output_path": "{{version._path}}/Output/{{version.take_name}}",
                "extra_folders": "",
                "environments": ["Maya", "Houdini", "Nuke", "Fusion", "Photoshop"],
                "type_for": "Asset"
            },
            {
                "name": "Previs",
                "code": "Previs",
                "path": "{{project.code}}/Sequences/{{sequence.code}}/Shots/{{version.base_name}}/{{type.code}}",
                "filename": "{{version.base_name}}_{{version.take_name}}_{{type.code}}_v{{'%03d'|format(version.version_number)}}_{{version.created_by.initials}}{{version.extension}}",
                "output_path": "{{version._path}}/Output/{{version.take_name}}",
                "extra_folders": "",
                "environments": ["Maya", "Houdini"],
                "type_for": "Shot"
            },
            {
                "name": "Lighting",
                "code": "Lighting",
                "path": "{{project.code}}/Sequences/{{sequence.code}}/Shots/{{version.base_name}}/{{type.code}}",
                "filename": "{{version.base_name}}_{{version.take_name}}_{{type.code}}_v{{'%03d'|format(version.version_number)}}_{{version.created_by.initials}}{{version.extension}}",
                "output_path": "{{version._path}}/Output/{{version.take_name}}",
                "extra_folders": "",
                "environments": ["Maya", "Houdini"],
                "type_for": "Shot"
            },
            {
                "name": "Rig",
                "code": "Rig",
                "path": "{{project.code}}/Assets/{{asset.type}}/{{version.base_name}}/{{type.code}}",
                "filename": "{{version.base_name}}_{{version.take_name}}_{{type.code}}_v{{'%03d'|format(version.version_number)}}_{{version.created_by.initials}}{{version.extension}}",
                "output_path": "{{version._path}}/Output/{{version.take_name}}",
                "extra_folders": "",
                "environments": ["Maya", "Houdini"],
                "type_for": "Asset"
            },
            {
                "name": "Roto",
                "code": "Roto",
                "path": "{{project.code}}/Sequences/{{sequence.code}}/Shots/{{version.base_name}}/{{type.code}}",
                "filename": "{{version.base_name}}_{{version.take_name}}_{{type.code}}_v{{'%03d'|format(version.version_number)}}_{{version.created_by.initials}}{{version.extension}}",
                "output_path": "{{version._path}}/Output/{{version.take_name}}",
                "extra_folders": "",
                "environments": ["Nuke", "Fusion"],
                "type_for": "Shot"
            },
            {
                "name": "Layout",
                "code": "Layout",
                "path":"{{project.code}}/Sequences/{{sequence.code}}/Shots/{{version.base_name}}/{{type.code}}",
                "filename": "{{version.base_name}}_{{version.take_name}}_{{type.code}}_v{{'%03d'|format(version.version_number)}}_{{version.created_by.initials}}{{version.extension}}",
                "output_path": "{{version._path}}/Output/{{version.take_name}}",
                "extra_folders": "",
                "environments": ["Maya", "Houdini"],
                "type_for": "Shot"
            },
            {
                "name": "Matte",
                "code": "Matte",
                "path": "{{project.code}}/Sequences/{{sequence.code}}/Shots/{{version.base_name}}/{{type.code}}",
                "filename": "{{version.base_name}}_{{version.take_name}}_{{type.code}}_v{{'%03d'|format(version.version_number)}}_{{version.created_by.initials}}{{version.extension}}",
                "output_path": "{{version._path}}/Output/{{version.take_name}}",
                "extra_folders": "",
                "environments": ["Photoshop"],
                "type_for": "Shot"
            },
            {
                "name": "Texture",
                "code": "Texture",
                "path": "{{project.code}}/Assets/{{asset.type}}/{{version.base_name}}/{{type.code}}",
                "filename": "{{version.base_name}}_{{version.take_name}}_{{type.code}}_v{{'%03d'|format(version.version_number)}}_{{version.created_by.initials}}{{version.extension}}",
                "output_path": "{{version._path}}/Output/{{version.take_name}}",
                "extra_folders": "",
                "environments": ["Photoshop", "Nuke", "Fusion"],
                "type_for": "Asset",
            },
            {
                "name": "Illustration",
                "code": "Illust",
                "path": "{{project.code}}/Assets/{{asset.type}}/{{version.base_name}}/{{type.code}}",
                "filename": "{{version.base_name}}_{{version.take_name}}_{{type.code}}_v{{'%03d'|format(version.version_number)}}_{{version.created_by.initials}}{{version.extension}}",
                "output_path": "{{version._path}}/Output/{{version.take_name}}",
                "extra_folders": "",
                "environments": ["Photoshop"],
                "type_for": "Asset"
            },
            {
                "name": "Look Development",
                "code": "LookDev",
                "path": "{{project.code}}/Assets/{{asset.type}}/{{version.base_name}}/{{type.code}}",
                "filename": "{{version.base_name}}_{{version.take_name}}_{{type.code}}_v{{'%03d'|format(version.version_number)}}_{{version.created_by.initials}}{{version.extension}}",
                "output_path": "{{version._path}}/Output/{{version.take_name}}",
                "extra_folders": "",
                "environments": ["Maya", "Houdini"],
                "type_for": "Asset"
            },
            {
                "name": "Match Move",
                "code": "MM",
                "path": "{{project.code}}/Sequences/{{sequence.code}}/Shots/{{version.base_name}}/{{type.code}}",
                "filename": "{{version.base_name}}_{{version.take_name}}_{{type.code}}_v{{'%03d'|format(version.version_number)}}_{{version.created_by.initials}}{{version.extension}}",
                "output_path": "{{version._path}}/Output/{{version.take_name}}",
                "extra_folders": "",
                "environments": ["3DEqualizer"],
                "type_for": "Shot"
            }
        ],
        
        maya_workspace_file_content = """workspace -fr "3dPaintTextures" ".maya/sourceimages/3dPaintTextures/";
workspace -fr "Adobe(R) Illustrator(R)" ".maya/data/";
workspace -fr "aliasWire" ".mayaFiles/data/";
workspace -fr "animImport" ".mayaFiles/data/";
workspace -fr "animExport" ".mayaFiles/data/";
workspace -fr "audio" ".mayaFiles/sound/";
workspace -fr "autoSave" ".mayaFiles/autosave/";
workspace -fr "clips" ".mayaFiles/clips/";
workspace -fr "DAE_FBX" ".mayaFiles/data/";
workspace -fr "DAE_FBX export" ".mayaFiles/data/";
workspace -fr "depth" ".mayaFiles/renderData/depth/";
workspace -fr "diskCache" ".mayaFiles/cache/";
workspace -fr "DXF" ".mayaFiles/data/";
workspace -fr "DXF export" ".mayaFiles/data/";
workspace -fr "DXF_FBX" ".mayaFiles/data/";
workspace -fr "DXF_FBX export" ".mayaFiles/data/";
workspace -fr "eps" ".mayaFiles/data/";
workspace -fr "EPS" ".mayaFiles/data/";
workspace -fr "FBX" ".mayaFiles/data/";
workspace -fr "FBX export" ".mayaFiles/data/";
workspace -fr "fluidCache" ".mayaFiles/cache/fluid/";
workspace -fr "furAttrMap" ".mayaFiles/renderData/fur/furAttrMap/";
workspace -fr "furEqualMap" ".mayaFiles/renderData/fur/furEqualMap/";
workspace -fr "furFiles" ".mayaFiles/renderData/fur/furFiles/";
workspace -fr "furImages" ".mayaFiles/renderData/fur/furImages/";
workspace -fr "furShadowMap" ".mayaFiles/renderData/fur/furShadowMap/";
workspace -fr "IGES" ".mayaFiles/data/";
workspace -fr "IGESexport" ".mayaFiles/data/";
workspace -fr "illustrator" ".mayaFiles/data/";
workspace -fr "image" ".mayaFiles/images/";
workspace -fr "images" ".mayaFiles/images/";
workspace -fr "iprImages" ".mayaFiles/renderData/iprImages/";
workspace -fr "lights" ".mayaFiles/renderData/shaders/";
workspace -fr "mayaAscii" ".mayaFiles/scenes/";
workspace -fr "mayaBinary" ".mayaFiles/scenes/";
workspace -fr "mel" ".mayaFiles/scripts/";
workspace -fr "mentalray" ".mayaFiles/renderData/mentalray/";
workspace -fr "mentalRay" ".mayaFiles/renderData/mentalray";
workspace -fr "move" ".mayaFiles/data/";
workspace -fr "movie" ".mayaFiles/movies/";
workspace -fr "OBJ" ".mayaFiles/data/";
workspace -fr "OBJexport" ".mayaFiles/data/";
workspace -fr "offlineEdit" ".mayaFiles/scenes/edits/";
workspace -fr "particles" ".mayaFiles/particles/";
workspace -fr "renderData" ".mayaFiles/renderData/";
workspace -fr "renderScenes" ".mayaFiles/scenes/";
workspace -fr "RIB" ".mayaFiles/data/";
workspace -fr "RIBexport" ".mayaFiles/data/";
workspace -fr "scene" ".mayaFiles/scenes/";
workspace -fr "scripts" ".mayaFiles/scripts/";
workspace -fr "shaders" ".mayaFiles/renderData/shaders/";
workspace -fr "sound" ".mayaFiles/sound/";
workspace -fr "sourceImages" ".mayaFiles/sourceimages/";
workspace -fr "templates" ".mayaFiles/assets/";
workspace -fr "textures" ".mayaFiles/images/";
workspace -fr "translatorData" ".mayaFiles/data/";
"""
    )
    
    def __init__(self):
        
        self.config_values = Config.default_config_values.copy()
        self.user_config = {}
        
        # the priority order is
        # oyProjectManager.config
        # config.py under .oyrc directory
        # config.py under $OYPROJECTMANAGER_PATH
        
        self._parse_settings()
    
    def _parse_settings(self):
        # for now just use $OYPROJECTMANAGER_PATH
        ENV_KEY = "OYPROJECTMANAGER_PATH"
        
        # try to get the environment variable
        if not os.environ.has_key(ENV_KEY):
            # don't do anything
            logger.debug("no environment key found for user settings")
        else:
            logger.debug("environment key found")
            
            resolved_path = os.path.expanduser(
                os.path.join(
                    os.environ[ENV_KEY],
                    "config.py"
                )
            )
            
            # using `while` is not safe to expand variables
            # do the expansion for 5 times which is complex enough
            # and I don't (hopefully) expect anybody to use
            # more than 5 level deep environment variables
            resolved_path = os.path.expandvars(
                os.path.expandvars(
                    os.path.expandvars(
                        os.path.expandvars(
                            resolved_path
                        )
                    )
                )
            )
            
            try:
                try:
                    logger.debug("importing user config")
                    execfile(resolved_path, self.user_config)
                except SyntaxError, err:
                    raise RuntimeError("There is a syntax error in your "
                                       "configuration file: " + str(err))
                
                # append the data to the current settings
                logger.debug("updating system config")
                for key in self.user_config:
                    if key in self.config_values:
                        self.config_values[key] = self.user_config[key]
            
            except IOError:
                logger.warning("The $OYPROJETMANAGER_PATH:" + resolved_path + \
                               " doesn't exists! skipping user config")
    
    def __getattr__(self, name):
        return self.config_values[name]
    
    def __getitem__(self, name):
        return getattr(self, name)
    
    def __setitem__(self, name, value):
        return setattr(self, name, value)
    
    def __delitem__(self, name):
        delattr(self, name)
    
    def __contains__(self, name):
        return name in self.config_values
    
    @property
    def last_user_id(self):
        """returns the last user id
        
        It is not very much related with the config.py and user settings, but
        it seems the most appropriate place is this one to get information from
        individual users.
        
        This should work fairly fast, because it uses the local filesystem not
        the network thus the fileserver.
        """
        # TODO: This should be replaced with beaker.session
        
        file_name = 'last_user_id'
        file_path = os.path.expanduser("~/.oypmrc/")
        file_full_path = os.path.join(file_path, file_name)
        
        last_user_id = None
        
        try:
            last_user_file = open(file_full_path)
        except IOError:
            pass
        else:
            last_user_id = int(last_user_file.readline().strip())
            last_user_file.close()
        
        return last_user_id
        
    @last_user_id.setter
    def last_user_id(self, user_id):
        """sets the user id for the last user
        """
        if not isinstance(user_id, int):
            raise RuntimeWarning("user_id for last_user_id should be an int")
            return
        
        file_name = 'last_user_id'
        file_path = os.path.expanduser("~/.oypmrc/")
        file_full_path = os.path.join(file_path, file_name)
        
        logger.debug("saving user id to %s" % file_full_path)
        
        # create the folder first
        try:
            os.makedirs(file_path)
        except OSError:
            # already created
            pass
        
        try:
            last_user_file = open(file_full_path, 'w')
        except IOError as e:
            pass
        else:
            last_user_file.write(str(user_id))
            last_user_file.close()
